bagua/torch_api/contrib/utils/redis_store.py

```diff
@@ -0,0 +1,231 @@
+import socket
+import subprocess
+import time
+from bagua.torch_api.env import get_rank, get_local_rank, get_world_size, get_local_size
+
+try:
+    from redis import Redis
+except ImportError:
+    print(
+        "DEBUG: did not find redis-py. To install it, run `pip install redis` or follow instructions on its website(https://github.com/andymccurdy/redis-py)."
+    )
+    raise
+
+from typing import List, Dict, Optional, Union
+from .store import Store, ClusterStore
+import torch.distributed.distributed_c10d as c10d
+import json
+import logging
+import atexit
+
+
+try:
+    p = subprocess.Popen(
+        ["redis-server", "--version"], stdout=subprocess.PIPE, stderr=subprocess.PIPE
+    )
+except Exception:
+    print(
+        "DEBUG: did not find redis-server. Follow instructions on its website(https://redis.io/download) to have it installed."
+    )
+    raise
+
+
+__all__ = ["RedisStore"]
+
+_host_ip = None
+
+_global_redis_servers = []
+
+
+class RedisStore(ClusterStore):
+    """
+    A Redis-based distributed Key-Value store implementation, with ``set(...)`` and ``get(...)``` API exposed.
+
+    Args:
+        hosts (List[Dict[str, str]]): A list of redis servers, defined by a list of dict containing Redis host and
+            port information like ``[{"host": "192.168.1.0", "port": "7000"}, {"host": "192.168.1.1", "port": "7000"}]``.
+            A new Redis instance will be spawned on each node if ``hosts=None``.
+        cluster_mode (bool): If ``True``, data is sharded across all Redis instances. Otherwise, if there are ``m``
+            Redis instances, the workers on the n-th node will use the ``n % m``-th Redis instance.
```

Comment:
check if these are correct descriptions

Reply:
Only one redis server can be bootstrapped on each node, thus the maximum memory limit of it is determined on its first initialization.

Reply:
added